namecache/namecache.go

```diff
@@ -0,0 +1,151 @@
+package namecache
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	merkledag "github.com/ipfs/go-ipfs/merkledag"
+	namesys "github.com/ipfs/go-ipfs/namesys"
+	path "github.com/ipfs/go-ipfs/path"
+	pin "github.com/ipfs/go-ipfs/pin"
+	uio "github.com/ipfs/go-ipfs/unixfs/io"
+
+	logging "gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log"
+)
+
+const (
+	followInterval = 60 * time.Minute
+)
+
+var log = logging.Logger("namecache")
+
+// NameCache represents a following cache of names
+type NameCache interface {
+	Follow(name string, pinit bool)
+	Unfollow(name string)
+	ListFollows() []string
+}
+
+type nameCache struct {
+	nsys    namesys.NameSystem
+	pinning pin.Pinner
+	dag     merkledag.DAGService
+
+	ctx     context.Context
+	follows map[string]func()
+	mx      sync.Mutex
+}
+
+func NewNameCache(ctx context.Context, nsys namesys.NameSystem, pinning pin.Pinner, dag merkledag.DAGService) NameCache {
+	return &nameCache{
+		ctx:     ctx,
+		nsys:    nsys,
+		pinning: pinning,
+		dag:     dag,
+		follows: make(map[string]func()),
+	}
+}
+
+// Follow spawns a goroutine that periodically resolves a name
+// and (when pinit is true) pins it in the background
+func (nc *nameCache) Follow(name string, pinit bool) {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	if _, ok := nc.follows[name]; ok {
+		return
+	}
+
+	ctx, cancel := context.WithCancel(nc.ctx)
+	go nc.followName(ctx, name, pinit)
+	nc.follows[name] = cancel
+}
+
+// Unfollow cancels a follow
+func (nc *nameCache) Unfollow(name string) {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	cancel, ok := nc.follows[name]
+	if ok {
+		cancel()
+		delete(nc.follows, name)
+	}
+}
+
+// ListFollows returns a list of names currently being followed
+func (nc *nameCache) ListFollows() []string {
+	nc.mx.Lock()
+	defer nc.mx.Unlock()
+
+	follows := make([]string, 0)
+	for name, _ := range nc.follows {
+		follows = append(follows, name)
+	}
+
+	return follows
+}
+
+func (nc *nameCache) followName(ctx context.Context, name string, pinit bool) {
+	nc.resolveAndPin(ctx, name, pinit)
+
+	ticker := time.NewTicker(followInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			nc.resolveAndPin(ctx, name, pinit)
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (nc *nameCache) resolveAndPin(ctx context.Context, name string, pinit bool) {
+	log.Debugf("resolving %s", name)
+
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	p, err := nc.nsys.Resolve(ctx, name)
+	if err != nil {
+		log.Debugf("error resolving %s: %s", name, err.Error())
+		return
+	}
+
+	log.Debugf("resolved %s to %s", name, p)
+
+	if !pinit {
+		return
+	}
+
+	log.Debugf("pinning %s", p)
+
+	r := &path.Resolver{
+		DAG:         nc.dag,
+		ResolveOnce: uio.ResolveUnixfsOnce,
+	}
+
+	n, err := r.ResolvePath(ctx, p)
+	if err != nil {
+		log.Debugf("error resolving path %s to node: %s", p, err.Error())
+		return
+	}
+
+	err = nc.pinning.Pin(ctx, n, true)
```

Comment:
Should call `pinning.Update` when the content wasn't pinned before previous resolve here.

Reply:
Oh and take the pinning lock somewhere around here - `defer n.Blockstore.PinLock().Unlock()`

Reply:
Oh yeah, the lock -- good point. 
Should `Update` be called instead of `Pin`?

Reply:
Actually seems like a more complicated logic is needed:
- In first resolution, call `Pin`.
- In subsequent resolutions:
  - Do nothing if the name resolves to the same node
  - Call `Update` if the pin has changed

I also wonder whether we should unpin when we cancel a follow.

Reply:
> I also wonder whether we should unpin when we cancel a follow.

I'd say yes. Also call `Pin` instead of `Update` when the hash was pinned before the follow, so we don't unintentionally unpin it.